crates/aptos-rest-client/src/lib.rs

```diff
@@ -1006,6 +1012,58 @@ impl Client {
         let (response, state) = self.check_response(response).await?;
         Ok(Response::new(response.bytes().await?, state))
     }
+
+    pub async fn try_until_ok<F, Fut, T>(
+        total_wait: Option<Duration>,
+        initial_interval: Option<Duration>,
+        function: F,
+    ) -> AptosResult<T>
+    where
+        F: Fn() -> Fut,
+        Fut: Future<Output = AptosResult<T>>,
+    {
+        let total_wait = total_wait.unwrap_or(DEFAULT_MAX_WAIT_DURATION);
+        let mut backoff = initial_interval.unwrap_or(DEFAULT_INTERVAL_DURATION);
+        let mut result = Err(RestError::Unknown(anyhow!("Failed to run function")));
+        let start = Instant::now();
+
+        // TODO: Add jitter
+        while start.elapsed() < total_wait {
+            result = function().await;
+
+            let retry = match &result {
+                Ok(_) => break,
+                Err(err) => match err {
+                    RestError::Api(inner) => is_retriable_status_code(inner.status_code.as_u16()),
+                    RestError::Http(inner) => is_retriable_status_code(*inner),
+                    RestError::Bcs(_)
+                    | RestError::Json(_)
+                    | RestError::Timeout(_)
+                    | RestError::Unknown(_) => true,
+                    RestError::UrlParse(_) => false,
+                },
+            };
+
+            if !retry {
+                break;
+            }
+
+            aptos_logger::info!(
+                "Failed to call API, retrying in {}ms: {:?}",
+                backoff.as_millis(),
+                result.as_ref().err().unwrap()
+            );
+
+            tokio::time::sleep(backoff).await;
+            backoff = backoff.saturating_mul(2);
+        }
+
+        result
+    }
+}
+
+fn is_retriable_status_code(status_code: u16) -> bool {
+    (500..600).contains(&status_code) || status_code == 404 || status_code == 429
```

Comment:
404 isn't retryable in many cases, if a resource doesn't exist sometimes it just doesn't exist, not because we're waiting for it to come into existence. We shouldn't retry blindly on all 404s.

Reply:
Can we make the codes we retry on configurable, with a sane default set as a constant that users could then modify when calling this.